tracer/src/Datadog.Tracer.Native/method_rewriter.cpp

```diff
@@ -516,26 +545,50 @@ HRESULT TracerMethodRewriter::Rewrite(RejitHandlerModule* moduleHandler, RejitHa
         reWriterWrapper.LoadLocalAddress(callTargetReturnIndex);
         tracerTokens->WriteCallTargetReturnGetReturnValue(&reWriterWrapper, callTargetReturnToken,
-                                                              &callTargetReturnGetReturnInstr);
+                                                          &callTargetReturnGetReturnInstr);
         reWriterWrapper.StLocal(returnValueIndex);
     }
 
     ILInstr* endMethodTryLeave = reWriterWrapper.CreateInstr(CEE_LEAVE_S);
 
-    // *** EndMethod call catch
+    ILInstr* filterEnd = nullptr;
     ILInstr* endMethodCatchFirstInstr = nullptr;
-    tracerTokens->WriteLogException(&reWriterWrapper, integration_type_ref, &caller->type,
-                                        &endMethodCatchFirstInstr);
+    if (corProfiler->call_target_bubble_up_exception_available)
+    {
+        Logger::Debug("Creating filter for try / catch for CallTargetBubbleUpException (end method).");
+        filterEnd = reWriterWrapper.CreateFilterForException(tracerTokens->GetExceptionTypeRef(),
+                                                             bubbleUpExceptionTypeRef,
+                                                             exceptionValueIndex);
+        endMethodCatchFirstInstr = reWriterWrapper.Pop();
+        reWriterWrapper.LoadLocal(exceptionValueIndex);
+    }
+
+    // transfer->m_pTarget = endFilter;
+    // *** EndMethod call catch
+    tracerTokens->WriteLogException(&reWriterWrapper, integration_type_ref, &caller->type, &endMethodCatchFirstInstr);
+
     ILInstr* endMethodCatchLeaveInstr = reWriterWrapper.CreateInstr(CEE_LEAVE_S);
 
     // *** EndMethod exception handling clause
     EHClause endMethodExClause{};
-    endMethodExClause.m_Flags = COR_ILEXCEPTION_CLAUSE_NONE;
-    endMethodExClause.m_pTryBegin = endMethodTryStartInstr;
-    endMethodExClause.m_pTryEnd = endMethodCatchFirstInstr;
-    endMethodExClause.m_pHandlerBegin = endMethodCatchFirstInstr;
-    endMethodExClause.m_pHandlerEnd = endMethodCatchLeaveInstr;
-    endMethodExClause.m_ClassToken = tracerTokens->GetExceptionTypeRef();
+    if (corProfiler->call_target_bubble_up_exception_available)
+    {
+        endMethodExClause.m_Flags = COR_ILEXCEPTION_CLAUSE_FILTER;
+        endMethodExClause.m_pTryBegin = endMethodTryStartInstr;
+        endMethodExClause.m_pTryEnd = filterEnd;
+        endMethodExClause.m_pHandlerBegin = endMethodCatchFirstInstr;
+        endMethodExClause.m_pHandlerEnd = endMethodCatchLeaveInstr;
+        endMethodExClause.m_pFilter = filterEnd;
+    }
+    else
+    {
+        endMethodExClause.m_Flags = COR_ILEXCEPTION_CLAUSE_NONE;
+        endMethodExClause.m_pTryBegin = endMethodTryStartInstr;
+        endMethodExClause.m_pTryEnd = endMethodCatchFirstInstr;
+        endMethodExClause.m_pHandlerBegin = endMethodCatchFirstInstr;
+        endMethodExClause.m_pHandlerEnd = endMethodCatchLeaveInstr;
+        endMethodExClause.m_ClassToken = tracerTokens->GetExceptionTypeRef();
+    }
```

Comment:
Do we want to change the Tracer instrumentation even if `AppSec` is disabled? 🤔 
In other words, will `CallTargetBubbleUpException` be in use regardless to AppSec?

Reply:
I think it might be good to have it regardless, for some other scenarios. And also, AppSec can be enabled via remote configuration without the app restarting.

Reply:
I would argue that having AppSec enabled via remote configuration would require ReJITting our integrations and reapply the instrumentation using the extra bits that are missing. Since your additional will only execute when an exception will be thrown, then I think it's fine because we can assume that it won't happen much. However, if you were adding a more impactful stream of IL that would execute every time and will be only relevant when certain product is enabled, then I would say that it should have been delayed until it's actually needed.

Reply:
good point, I'll resolve for now but will add a task in our backlog!